File: server/typescript/packages/render/src/extract/extract-map.ts
// Null-safe coercions from a ExtractionOutcome data map onto typed values. Generated
// extract(...) calls these. Mirrors Java ExtractMap.
//
// Tier-2 divergence: JS has one number type, so asInt/asLong both return `number | null`
// and truncate toward zero via Math.trunc (Java intValue()/longValue() also truncate).
// The numeric helpers gate on actual numbers (mirroring Java `instanceof Number`): a
// non-numeric string or a boolean returns null rather than coercing.

export function asString(d: Record<string, unknown>, k: string): string | null {
  const v = d[k];
  if (v == null) return null;
  return typeof v === "string" ? v : String(v);
}

export function asInt(d: Record<string, unknown>, k: string): number | null {
  const v = d[k];
  return typeof v === "number" && Number.isFinite(v) ? Math.trunc(v) : null;
}

export function asLong(d: Record<string, unknown>, k: string): number | null {
  const v = d[k];
  return typeof v === "number" && Number.isFinite(v) ? Math.trunc(v) : null;
}

export function asDouble(d: Record<string, unknown>, k: string): number | null {
  const v = d[k];
  return typeof v === "number" && Number.isFinite(v) ? v : null;
}

export function asBool(d: Record<string, unknown>, k: string): boolean | null {
  const v = d[k];
  return typeof v === "boolean" ? v : null;
}

export function asStringList(d: Record<string, unknown>, k: string): (string | null)[] | null {
  const v = d[k];
  if (!Array.isArray(v)) return null;
  return v.map((e) => (e == null ? null : typeof e === "string" ? e : String(e)));
}

File: server/java/render/src/main/java/com/metaobjects/render/extract/ExtractMap.java
package com.metaobjects.render.extract;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Null-safe coercions from a ExtractionOutcome data map onto typed record components. Generated extractLenient(...) calls these. */
public final class ExtractMap {
    private ExtractMap() {}

    public static String asString(Map<String, Object> d, String k) {
        Object v = d.get(k);
        return v == null ? null : (v instanceof String s ? s : String.valueOf(v));
    }

    public static Integer asInt(Map<String, Object> d, String k) {
        Object v = d.get(k);
        return v instanceof Number n ? n.intValue() : null;
    }

    public static Long asLong(Map<String, Object> d, String k) {
        Object v = d.get(k);
        return v instanceof Number n ? n.longValue() : null;
    }

    public static Double asDouble(Map<String, Object> d, String k) {
        Object v = d.get(k);
        return v instanceof Number n ? n.doubleValue() : null;
    }

    public static Boolean asBool(Map<String, Object> d, String k) {
        Object v = d.get(k);
        return v instanceof Boolean b ? b : null;
    }

    public static List<String> asStringList(Map<String, Object> d, String k) {
        Object v = d.get(k);
        if (!(v instanceof List<?> list)) return null;
        List<String> out = new ArrayList<>(list.size());
        for (Object e : list) out.add(e == null ? null : String.valueOf(e));
        return out;
    }
}
